modules/rostests/apitests/shell32/SHLimitInputEdit.cpp

```diff
@@ -0,0 +1,332 @@
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP GetDisplayNameOf(
+        PCUITEMID_CHILD pidl,
+        SHGDNF uFlags,
+        STRRET *pName)
+    {
+        return E_NOTIMPL;
+    }
+
+    STDMETHODIMP SetNameOf(
+        HWND hwnd,
+        PCUITEMID_CHILD pidl,
+        LPCWSTR pszName,
+        SHGDNF uFlags,
+        PITEMID_CHILD *ppidlOut)
+    {
+        return E_NOTIMPL;
+    }
+
+    /*** IItemNameLimits methods ***/
+
+    STDMETHODIMP GetMaxLength(
+        LPCWSTR pszName,
+        int *piMaxNameLen);
+
+    STDMETHODIMP GetValidCharacters(
+        LPWSTR *ppwszValidChars,
+        LPWSTR *ppwszInvalidChars);
+
+protected:
+    LONG m_nRefCount;
+    INT m_iMaxNameLen;
+    BOOL m_bDisabled;
+};
+
+STDMETHODIMP CShellFolder::QueryInterface(REFIID riid, void **ppvObject)
+{
+    trace("QueryInterface\n");
+
+    if (!ppvObject)
+        return E_POINTER;
+
+    if (IsEqualIID(riid, IID_IItemNameLimits))
+    {
+        trace("IID_IItemNameLimits\n");
+        *ppvObject = static_cast<IItemNameLimits *>(this);
+        return S_OK;
+    }
+    if (IsEqualIID(riid, IID_IShellFolder))
+    {
+        trace("IID_IShellFolder\n");
+        *ppvObject = static_cast<IShellFolder *>(this);
+        return S_OK;
+    }
+    if (IsEqualIID(riid, IID_IUnknown))
+    {
+        trace("IID_IUnknown\n");
+        *ppvObject = static_cast<IShellFolder *>(this);
+        return S_OK;
+    }
+
+    return E_NOINTERFACE;
+}
+
+STDMETHODIMP_(ULONG) CShellFolder::AddRef()
+{
+    trace("AddRef\n");
+    return ++m_nRefCount;
+}
+
+STDMETHODIMP_(ULONG) CShellFolder::Release()
+{
+    trace("Release\n");
+    if (--m_nRefCount == 0)
+    {
+        delete this;
+        return 0;
+    }
+    return m_nRefCount;
+}
+
+STDMETHODIMP CShellFolder::GetMaxLength(
+    LPCWSTR pszName,
+    int *piMaxNameLen)
+{
+    trace("GetMaxLength('%S', %p (%d))\n", pszName, piMaxNameLen, *piMaxNameLen);
+
+    if (!piMaxNameLen)
+        return E_POINTER;
+
+    *piMaxNameLen = m_iMaxNameLen;
+    return S_OK;
+}
+
+STDMETHODIMP CShellFolder::GetValidCharacters(
+    LPWSTR *ppwszValidChars,
+    LPWSTR *ppwszInvalidChars)
+{
+    trace("GetValidCharacters(%p, %p)\n", ppwszValidChars, ppwszInvalidChars);
+
+    if (m_bDisabled)
+        return E_NOTIMPL;
+
+    if (ppwszInvalidChars)
+    {
+        SHStrDupW(L"/:*?\"<>|", ppwszInvalidChars);
+    }
+    if (ppwszValidChars)
+    {
+        *ppwszValidChars = NULL;
+    }
+    return S_OK;
+}
+
+static INT_PTR CALLBACK
+DialogProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    if (uMsg != WM_INITDIALOG)
+        return FALSE;
+
+    HWND hEdt1 = GetDlgItem(hwnd, edt1);
+    INT n;
+    HRESULT hr;
+    WCHAR szText[64];
+
+    _SEH2_TRY
+    {
+        hr = SHLimitInputEdit(hEdt1, NULL);
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        hr = 0xDEAD;
+    }
+    _SEH2_END;
+    ok_int(hr, 0xDEAD);
+
+    CComPtr<IShellFolder> psf(new CShellFolder(123, FALSE));
```

Comment:
However, none of that reference counting really matters for the test case?

You can just create the object here like this:
`CShellFolder psf(123, FALSE);`

And just update the `AddRef` to always return 2, `Release` to always return 1.

Reply:
OK.